### ISSUE TITLE
Elif ladder not narrowing type consistently.

### ISSUE THREAD

**Issue by mezuzza** [microsoft/pyright#4207]
**Describe the bug**
Consider the following:
```python
def test(a: int | None, b: int | None):
  if a is None and b is None:
    # branch 1
    a = 1
    b = 2
  elif b is None:
    # branch 2
    b = 3
    a + b  # a is int | None
  elif a is None:
    # branch 3
    a = 4
    a + b
  else:
    # branch 4
    a + b
  return a + b  # a is int | None
```
In branch 2, `a` is typed as `int | None` even though it can be known as not None.
In branch 3, `b` is typed as `int` even though `a` wasn't in branch 2.
In the return statement, `a` is still `int | None` while it could be known to be `int` by the end of the `if` statement.

It seems as though branch 2 and branch 3 should agree that `a` and `b` respectively should not be None. 

**To Reproduce**
Run pyright on the function above

**Expected behavior**
The above code should type check without errors though warnings would still apply for unused expressions.

**VS Code extension or command-line**
Command line

**Additional context**
```sh
» python --version
Python 3.10.8
» pyright --version
pyright 1.1.280
```

My guess is that branch 3 is somehow using branch 2's check to prove that b is not None, but branch 2 isn't able to use branch 1 to do the same. I can see how this could be "working as intended", but it feels pretty strange.

**Comment by erictraut** [comment]
Types are narrowed for individual expressions (e.g. `a` or `b`) independently. There is no form of "conditional types" or "entaglement" between two expressions such as "`a` is `None` if `b` is `None` but is otherwise `int | None`". This might be theoretically possible, but in practice it would be computationally very expensive and complex. I'm not aware of any static type analyzer or compiler in any language that performs type narrowing like this.

In your example above, type narrowing is working consistently throughout if you consider `a` and `b` as independent expressions.

**Comment by mezuzza** [comment]
Fair enough.

The approach I thought of would be quite complex - though I'm not 100% sure it would be any more complex than traversing the AST of the `elif` ladder once.

That said, I don't know the internals of pyright, so I'll just trust that this is too hard.
